### PR TITLE
ROX-14150: Add config maps as input / output for declarative-config commands

### PR DESCRIPTION
**Pull request by dhaus67**
## Description

This is the third PR in a series of PRs to add supportive `roxctl` commands for the declarative config feature.

This PR adds the ability for the `create` and `lint` commands to accept config maps as output / input respectively.

Within the `create` command, clients have the possiblity to specify a config map to store the configuration. The YAML output will be stored under the key `<type of resource>-<name of resource>`.
In case the config map is not given, the command behaves as-is: writing the YAML output to STDOUT.

Within the `lint` command, clients have the possiblity to specify a config map to read the configuration from. Each key within the config map will be read and its assumed that each key contains a valid declarative configuration YAML.
In case the config map is not given, the command behaves as-is: expecting a file input that contains the YAML configuration.

Besides adjusting both commands and adding the possiblity to read / write from config maps, the following has been done:
- Created a util to hide inherited flags (a.k.a `PersistentFlags()`). 

The reason for this is that the by default created persistent flags contain information about central connectivity (endpoint, grpc type etc.), which may become confusing for clients using the declarative config subtree of commands since central connectivity is not required within those.

Additionally, it lead to some of the flags to be less discoverable that are defined as persistent flags (i.e. the config map flags, the shared flags for auth provider types). Hence, decided to hide all the inherited flags from the `declarative-config` subtree.

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed

- see unit tests.

### CODE REVIEW

**Comment by dhaus67** [comment]
Current dependencies on/for this PR:
* master
  * **PR #5415** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/5415" target="_blank">Graphite</a> 
    * **PR #5427** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/5427" target="_blank">Graphite</a> 
      * **PR #5462** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/5462" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/stackrox/stackrox/5462?utm_source=stack-comment).

File: roxctl/declarativeconfig/lint/lint.go [hunk 1]
```diff
@@ -30,31 +31,57 @@ func Command(cliEnvironment environment.Environment) *cobra.Command {
...
+
 	return cmd
 }
 
 type lintCmd struct {
 	env environment.Environment
 
 	file         string
-	fileContents []byte
+	fileContents [][]byte
```
Comment (janisz) [inline roxctl/declarativeconfig/lint/lint.go:50]: why this change? Do we allow content of multiple files?

File: roxctl/declarativeconfig/configmap/configmap.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+// ReadFromConfigMap will read all keys from a config map and return them in byte format.
+// The Kubernetes context will be inferred from either the $KUBECONFIG variable or the $HOME/.kube/config path.
+// Note: the binary data within the config map will be skipped, as the output shall be UTF-8 compatible.
+func ReadFromConfigMap(configMap, namespace string) ([][]byte, error) {
+	client, namespaceFromConfig, err := createK8SClient(namespace)
+	if err != nil {
+		return nil, err
+	}
+
+	cm, err := client.CoreV1().ConfigMaps(namespaceFromConfig).Get(context.TODO(), configMap, metav1.GetOptions{})
```
Comment (janisz) [inline roxctl/declarativeconfig/configmap/configmap.go:53]: why not passing context to this function?

File: roxctl/common/flags/hide.go [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+import (
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+	"github.com/stackrox/rox/pkg/set"
+)
+
+// HideInheritedFlags hides all inherited flags except the flag names given.
+// This is especially useful for commands that do not require central connectivity and want to avoid exposing these
+// flags to the client to avoid confusion.
+func HideInheritedFlags(cmd *cobra.Command, flagsToShow ...string) {
```
Comment (janisz) [inline roxctl/common/flags/hide.go:12]: ❤️ 

How about extracting this to a separated PR and add to 
- `version`
- `help`
- `completion`

File: roxctl/declarativeconfig/configmap/configmap.go [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+	if err != nil {
+		return nil, errors.Wrapf(err, "retrieving config map %s", configMap)
+	}
+	var contents [][]byte
```
Comment (janisz) [inline roxctl/declarativeconfig/configmap/configmap.go:57]: super nit: prealloc? 

Maybe we should enable linter for that although I think in CLI it makes little sense to optimize this

File: roxctl/declarativeconfig/configmap/configmap.go [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+	cfg, err := rawConfig.ClientConfig()
+	if err != nil {
+		return nil, "", errors.Wrap(err, "retrieving kubeconfig")
+	}
+
+	client, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		return nil, "", errors.Wrap(err, "creating kubernetes client")
+	}
+	if namespace == "" {
```
Comment (janisz) [inline roxctl/declarativeconfig/configmap/configmap.go:75]: invert if to make it easier to read and require less indents

File: roxctl/common/flags/hide.go (hunk 3 above)
Comment (dhaus67) [reply roxctl/common/flags/hide.go]: Help won't sadly work since its a built-in command by cobra :/

Nevertheless, created the follow-up PR: #5476

File: roxctl/declarativeconfig/lint/lint.go (hunk 1 above)
Comment (dhaus67) [reply roxctl/declarativeconfig/lint/lint.go:49]: Change was mostly done to only have a single variable. The config map can have multiple file contents, while the `--file` flag will only yield one. So, essentially this was done to avoid having one for the file input, and one for the config map input.

File: roxctl/declarativeconfig/lint/lint.go [hunk 6]
```diff
@@ -31,31 +32,55 @@ func Command(cliEnvironment environment.Environment) *cobra.Command {
...
+	if err != nil {
+		return errors.Wrap(err, "reading config map flag values")
+	}
+	l.configMap = configMap
+	l.namespace = namespace
+
+	if l.configMap != "" {
+		contents, err := configmap.ReadFromConfigMap(context.Background(), l.configMap, l.namespace)
+		if err != nil {
+			return errors.Wrapf(err, "reading from config map %s", l.configMap)
```
Comment (ivan-degtiarenko) [inline roxctl/declarativeconfig/lint/lint.go:66]:
```suggestion
			return errors.Wrapf(err, "reading from config map %s/%s", l.namespace, l.configMap)
```

File: roxctl/declarativeconfig/lint/lint.go [hunk 7]
```diff
@@ -31,31 +32,55 @@ func Command(cliEnvironment environment.Environment) *cobra.Command {
...
+func (l *lintCmd) Construct(cmd *cobra.Command) error {
+	configMap, namespace, err := configmap.ReadConfigMapFlags(cmd)
+	if err != nil {
+		return errors.Wrap(err, "reading config map flag values")
+	}
+	l.configMap = configMap
+	l.namespace = namespace
+
+	if l.configMap != "" {
+		contents, err := configmap.ReadFromConfigMap(context.Background(), l.configMap, l.namespace)
```
Comment (ivan-degtiarenko) [inline roxctl/declarativeconfig/lint/lint.go:64]: Should we put timeout on the context here?

File: go.mod [hunk 8]
```diff
@@ -349,7 +349,7 @@ require (
 	gopkg.in/warnings.v0 v0.1.2 // indirect
 	gopkg.in/yaml.v2 v2.4.0 // indirect
 	k8s.io/apiextensions-apiserver v0.26.1 // indirect
-	k8s.io/cli-runtime v0.26.0 // indirect
```
Comment (ivan-degtiarenko) [inline go.mod:352]: Should this line then move to a `require` section above?

**Review by ivan-degtiarenko — COMMENTED** [review]
Two big questions:
* do we need to introduce separate flags to handle secrets instead of config maps?
* can we add unit tests for that functionality?

File: roxctl/declarativeconfig/lint/lint.go (hunk 7 above)
Comment (dhaus67) [reply roxctl/declarativeconfig/lint/lint.go:64]: I'll have the timeout within `ReadConfigMap`, that makes things a lot easier for clients.

File: roxctl/declarativeconfig/lint/lint.go (hunk 6 above)
Comment (dhaus67) [reply roxctl/declarativeconfig/lint/lint.go:66]: Moved this error message with the namespace to `ReadConfigMap` instead.

File: roxctl/declarativeconfig/create/create.go [hunk 9]
```diff
@@ -18,5 +19,16 @@ func Command(cliEnvironment environment.Environment) *cobra.Command {
 		permissionSetCommand(cliEnvironment),
 		roleCommand(cliEnvironment),
 	)
+
+	c.PersistentFlags().String(k8sobject.ConfigMapFlag, "", `Config Map to which the declarative config YAML should be written to.
+If left empty, the created YAML will be printed to stdout instead`)
```
Comment (ivan-degtiarenko) [inline roxctl/declarativeconfig/create/create.go:24]:
```suggestion
If both this flag and  "secret" flag values are left empty, the created YAML will be printed to stdout instead`)
```

File: roxctl/declarativeconfig/create/create.go [hunk 10]
```diff
@@ -18,5 +19,16 @@ func Command(cliEnvironment environment.Environment) *cobra.Command {
...
+	c.PersistentFlags().String(k8sobject.SecretFlag, "", `Secret to which the declarative config YAML should be written to.
+Secrets should be used in case sensitive data is contained within the declarative configuration, e.g. within auth providers.
+If left empty, the created YAML will be printed to stdout instead.`)
```
Comment (ivan-degtiarenko) [inline roxctl/declarativeconfig/create/create.go:27]:
```suggestion
If both this flag and  "config-map" flag values are left empty, the created YAML will be printed to stdout instead`)
```

File: roxctl/declarativeconfig/k8sobject/k8sobject.go [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+		return err
+	}
+
+	if configMap != "" {
+		return writeConfigMap(k8sCtx, client, configMap, namespaceFromConfig, key, yaml)
+	}
+	return writeSecret(k8sCtx, client, secret, namespaceFromConfig, key, yaml)
+}
+
+func writeConfigMap(ctx context.Context, client kubernetes.Interface, configMap, namespace, key string, yaml []byte) error {
```
Comment (ivan-degtiarenko) [inline roxctl/declarativeconfig/k8sobject/k8sobject.go:40]: nit: `configMapName`

File: roxctl/declarativeconfig/lint/lint.go [hunk 12]
```diff
@@ -31,26 +36,53 @@ func Command(cliEnvironment environment.Environment) *cobra.Command {
 
 	cmd.Flags().StringVarP(&lintCmd.file, "file", "f", "", "file containing the declarative configuration in YAML format")
 
-	utils.Must(cmd.MarkFlagRequired("file"))
+	cmd.Flags().String(k8sobject.ConfigMapFlag, "", fmt.Sprintf(k8sObjectFlagTemplate, "config map"))
+	cmd.Flags().String(k8sobject.SecretFlag, "", fmt.Sprintf(k8sObjectFlagTemplate, "secret"))
+	cmd.Flags().String(k8sobject.NamespaceFlag, "", `namespace of the config map from which to read the declarative configuration from.
+In case this is not set, the namespace set within the current kube config context will be used`)
+
+	cmd.MarkFlagsMutuallyExclusive("file", k8sobject.ConfigMapFlag)
```
Comment (ivan-degtiarenko) [inline roxctl/declarativeconfig/lint/lint.go:44]: also `secret` and `file` are mutually exclusive

File: roxctl/declarativeconfig/lint/lint.go [hunk 13]
```diff
@@ -31,26 +36,53 @@ func Command(cliEnvironment environment.Environment) *cobra.Command {
...
+func (l *lintCmd) Construct(cmd *cobra.Command) error {
+	configMap, secret, namespace, err := k8sobject.ReadK8sObjectFlags(cmd)
+	if err != nil {
+		return errors.Wrap(err, "reading config map flag values")
+	}
+	l.configMap = configMap
+	l.secret = secret
+	l.namespace = namespace
+
+	if l.configMap != "" {
```
Comment (ivan-degtiarenko) [inline roxctl/declarativeconfig/lint/lint.go:70]:
```suggestion
	if l.configMap != "" || l.secret != "" {
```

File: roxctl/declarativeconfig/create/access_scope.go [hunk 14]
```diff
@@ -108,8 +130,16 @@ func (a *accessScopeCmd) PrintYAML() error {
 	if err := lint.Lint(yamlOut.Bytes()); err != nil {
 		return errors.Wrap(err, "linting the YAML output")
 	}
-	_, err := a.env.InputOutput().Out().Write(yamlOut.Bytes())
-	return errors.Wrap(err, "writing the YAML output")
+	if a.writeToK8sObject {
```
Comment (ivan-degtiarenko) [inline roxctl/declarativeconfig/create/access_scope.go:133]: nit: I don't see much profit in introducing such field since it is used only once

**Review by ivan-degtiarenko — APPROVED** [review]
Some nits, otherwise LGTM
